File: mobile/ios/Runner/NativeShell/ShellBridge.swift
import Flutter
import UIKit

final class ShellBridge {
  static let shared = ShellBridge()

  enum AuthState {
    case unknown
    case signedOut
    case signedIn
  }

  private(set) var authState: AuthState = .unknown
  var onAuthStateChange: ((AuthState) -> Void)?

  /// Declared by Dart: order is identity across the channel, and labels are localised.
  private(set) var tabs: [ShellTab] = []

  private(set) var searchPlaceholder: String?
  var onSearchPlaceholderChange: ((String?) -> Void)?
  var onSearchTextChange: ((String) -> Void)?

  var channel: FlutterMethodChannel?
  private var dartIsReady = false

  var isReady: Bool { dartIsReady }
  private var pendingRoute: String?
  private var pendingSettle: ((String) -> Void)?

  func attach(to engine: FlutterEngine) {
    let channel = FlutterMethodChannel(name: "immich/shell", binaryMessenger: engine.binaryMessenger)
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call)
      result(nil)
    }
    self.channel = channel
  }

  private func handle(_ call: FlutterMethodCall) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "ready":
      dartIsReady = true
      tabs = (args["tabs"] as? [[String: Any]] ?? []).compactMap(ShellTab.init)
      shellLog("[shell] dart ready, tabs=[%@]", tabs.map(\.id).joined(separator: ","))
      if let route = pendingRoute {
        pendingRoute = nil
        let settle = pendingSettle ?? { _ in }
        pendingSettle = nil
        send(route: route, whenSettled: settle)
      }
      scheduleDebugHooks()
    case "auth":
      let signedIn = args["signedIn"] as? Bool ?? false
      let next: AuthState = signedIn ? .signedIn : .signedOut
      guard next != authState else { return }
      authState = next
      shellLog("[shell] auth signedIn=%@", signedIn ? "yes" : "no")
      onAuthStateChange?(next)
    case "route":
      shellLog("[shell] dart route=%@", args["name"] as? String ?? "?")
    case "openViewer":
      openViewer(
        session: args["session"] as? Int ?? TimelineSessions.mainSession,
        index: args["index"] as? Int ?? 0
      )
    case "log":
      shellLog("[shell:dart] %@", args["text"] as? String ?? "?")
    case "search":
      if let placeholder = args["placeholder"] as? String {
        searchPlaceholder = placeholder
        onSearchPlaceholderChange?(placeholder)
      }
      if let text = args["text"] as? String {
        onSearchTextChange?(text)
      }
    case "barCollapsed":
      let route = args["route"] as? String ?? ""
      let collapsed = args["collapsed"] as? Bool ?? false
      frame(named: route)?.setCollapsed(collapsed)
    case "sync":
      let routes = (args["routes"] as? [[String: Any]] ?? []).map {
        MirrorFrame(
          name: $0["name"] as? String ?? "?",
          title: $0["title"] as? String,
          actions: $0["actions"] as? [[String: Any]] ?? [],
          hero: $0["hero"] as? Bool ?? false
        )
      }
      let tab = args["tab"] as? String ?? ""
      if args["claimTab"] as? Bool ?? false {
        selectIfNeeded(tab: tab)
      }
      applyTabBars(args["tabBars"] as? [String: [String: Any]] ?? [:])
      reconcile(to: routes, tab: tab)
      ShellEngine.shared.dartIsShowing(
        args["surface"] as? String ?? "",
        overlay: args["overlay"] as? Bool ?? false
      )
    default:
      shellLog("[shell] unhandled dart call %@", call.method)
    }
  }


  private func openViewer(session: Int, index: Int) {
    guard let nav = activeNavigationController else {
      shellLog("[shell:nav] viewer at %ld with no native stack to put it in", index)
      return
    }
    guard let source = TimelineSessions.shared.source(for: session) else { return }
    let viewer = AssetViewerController(source: source, startIndex: index)
    viewer.onClosed = {
      TimelineSessions.shared.close(session: session)
    }
    shellLog("[shell:nav] native viewer at %ld on session %ld over %@", index, session, describe(nav))
    nav.pushViewController(viewer, animated: true)
  }


  /// Setting `UITabBarController.tabs` retires `viewControllers` and `selectedIndex`,
  /// so every lookup goes through the shell rather than the tab bar itself.
  private var shell: NativeShellController? {
    let root = UIApplication.shared.connectedScenes
      .compactMap { ($0 as? UIWindowScene)?.keyWindow?.rootViewController }
      .first
    return root as? NativeShellController
  }

  var activeNavigationController: UINavigationController? { shell?.activeNavigationController }

  private func selectIfNeeded(tab: String) {
    guard let shell, shell.selectedTabId != tab else { return }
    shellLog("[shell:nav] dart moved to %@", tab)
    shell.select(tabId: tab)
  }

  /// A tab's root is not a stack frame, so its bar arrives separately.
  private func applyTabBars(_ bars: [String: [String: Any]]) {
    for (tab, bar) in bars {
      guard let host = navigationController(forTab: tab)?.viewControllers.first as? FlutterTabController else {
        continue
      }
      host.barRoute = bar["name"] as? String ?? tab
      host.apply(title: bar["title"] as? String, actions: bar["actions"] as? [[String: Any]] ?? [])
    }
  }

  private func navigationController(forTab tab: String) -> UINavigationController? {
    shell?.navigationController(forTab: tab)
  }

  private func frame(named route: String) -> FlutterStackController? {
    activeNavigationController?.viewControllers
      .compactMap { $0 as? FlutterStackController }
      .first { $0.shellLabel == route }
  }

  struct MirrorFrame {
    let name: String
    let title: String?
    let actions: [[String: Any]]
    let hero: Bool
  }

  private func reconcile(to frames: [MirrorFrame], tab: String) {
    guard let nav = navigationController(forTab: tab) else {
      shellLog("[shell:nav] sync [%@] with no native stack to put it in", frames.map(\.name).joined(separator: ","))
      return
    }
    let current = nav.viewControllers
    if (current.last as? AssetViewerController) != nil {
      shellLog("[shell:nav] sync [%@] under the native viewer, ignored", frames.map(\.name).joined(separator: ","))
      return
    }
    let base = current.prefix { !($0 is FlutterStackController) }
    let existing = current.compactMap { $0 as? FlutterStackController }

    var shared = 0
    while shared < existing.count, shared < frames.count, existing[shared].shellLabel == frames[shared].name {
      shared += 1
    }

    for (index, frame) in frames.prefix(shared).enumerated() {
      existing[index].apply(title: frame.title, actions: frame.actions, hero: frame.hero)
    }

    guard shared != existing.count || shared != frames.count else { return }

    for frame in existing.dropFirst(shared) {
      frame.suppressDartPop = true
    }

    var target: [UIViewController] = Array(base) + Array(existing.prefix(shared))
    for frame in frames.dropFirst(shared) {
      let controller = FlutterStackController(label: frame.name, nativeTitle: frame.title)
      controller.apply(title: frame.title, actions: frame.actions, hero: frame.hero)
      target.append(controller)
    }

    shellLog(
      "[shell:nav] sync %@ -> [%@]",
      describe(nav),
      target.map { ($0 as? ShellFlutterHost)?.shellLabel ?? "?" }.joined(separator: ",")
    )

    if target.count == current.count + 1, shared == existing.count, let pushed = target.last {
      nav.pushViewController(pushed, animated: true) // One added on top.
    } else if target.count < current.count, shared == frames.count {
      nav.setViewControllers(target, animated: true) // Removed from the top.
    } else {
      nav.setViewControllers(target, animated: false)
    }

    if target.count > current.count {
      scheduleDebugPop(on: nav)
    }
  }

  func describe(_ nav: UINavigationController) -> String {
    let labels = nav.viewControllers.map { ($0 as? ShellFlutterHost)?.shellLabel ?? String(describing: type(of: $0)) }
    return "[\(labels.joined(separator: ","))]"
  }

  func requestSync() {
    channel?.invokeMethod("resync", arguments: nil)
  }

  func requestDartPop(route: String) {
    channel?.invokeMethod("popFromNative", arguments: ["name": route])
  }

  func barAction(route: String, index: Int, item: Int) {
    channel?.invokeMethod("barAction", arguments: ["route": route, "index": index, "item": item])
  }

  func submitSearch(_ text: String) {
    shellLog("[shell] search submit %@", text.isEmpty ? "(cleared)" : text)
    channel?.invokeMethod("searchSubmitted", arguments: ["text": text])
  }

  func popToRoot(tab: String) {
    channel?.invokeMethod("popToRoot", arguments: ["tab": tab])
  }


  private var reportedInsets: UIEdgeInsets?

  func report(insets: UIEdgeInsets) {
    guard reportedInsets != insets else { return }
    reportedInsets = insets
    channel?.invokeMethod(
      "insets",
      arguments: [
        "top": Double(insets.top),
        "bottom": Double(insets.bottom),
        "left": Double(insets.left),
        "right": Double(insets.right),
      ]
    )
    shellLog("[shell] insets top=%.0f bottom=%.0f", insets.top, insets.bottom)
  }

  func show(route: String, whenSettled: @escaping (String) -> Void) {
    guard dartIsReady else {
      pendingRoute = route
      pendingSettle = whenSettled
      return
    }
    send(route: route, whenSettled: whenSettled)
  }

  private func send(route: String, whenSettled: @escaping (String) -> Void) {
    shellLog("[shell] show route=%@", route)
    guard let channel else {
      whenSettled("")
      return
    }
    channel.invokeMethod("show", arguments: ["route": route]) { reply in
      whenSettled((reply as? [String: Any])?["surface"] as? String ?? "")
    }
  }

  func capture(_ completion: @escaping (UIImage?) -> Void) {
    guard dartIsReady, let channel else {
      completion(nil)
      return
    }
    channel.invokeMethod("capture", arguments: nil) { reply in
      guard let data = reply as? FlutterStandardTypedData else {
        completion(nil)
        return
      }
      completion(UIImage(data: data.data, scale: 1))
    }
  }
}

File: mobile/android/app/src/main/kotlin/app/alextran/immich/shell/ShellBridge.kt
package app.alextran.immich.shell

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** The native half of `immich/shell`. See `hybrid-shell-protocol.md`. */
object ShellBridge {
  enum class AuthState { UNKNOWN, SIGNED_OUT, SIGNED_IN }

  /** Compose state, so the root swaps itself when these change. */
  var authState by mutableStateOf(AuthState.UNKNOWN)
    private set

  /** Declared by Dart: order is identity across the channel, and labels are localised. */
  var tabs by mutableStateOf<List<ShellTab>>(emptyList())
    private set

  /** The tab the native bar shows. Written by a tap here or by a `claimTab` sync from Dart. */
  var selectedTab by mutableStateOf("")

  var searchPlaceholder by mutableStateOf<String?>(null)
    private set

  /** Dart may set the field's text (a suggestion picked, a cleared query); the field owns edits. */
  var searchText by mutableStateOf("")

  /** The palette Flutter is wearing, so the chrome above it matches; null until Dart says. */
  var palette by mutableStateOf<ShellPalette?>(null)
    private set

  data class MirrorFrame(
    val name: String,
    val title: String?,
    val actions: List<Map<String, Any?>>,
    val hero: Boolean,
  )

  /** Stack reconciliation lives with whoever owns the stacks; the bridge only parses. */
  var onSync: ((frames: List<MirrorFrame>, tabBars: Map<String, MirrorFrame>, tab: String) -> Unit)? = null
  var onBarCollapsed: ((route: String, collapsed: Boolean) -> Unit)? = null
  var onBarScroll: ((route: String, progress: Float) -> Unit)? = null
  /** Dart is about to paint a different route; whatever the surface shows now is about to go. */
  var onWillChange: (() -> Unit)? = null
  var onTabsDeclared: ((List<ShellTab>) -> Unit)? = null
  var onOpenViewer: ((session: Int, index: Int) -> Unit)? = null

  private var channel: MethodChannel? = null
  var isReady = false
    private set
  private var pendingRoute: String? = null
  private var pendingSettle: ((String) -> Unit)? = null

  fun attach(engine: FlutterEngine) {
    val channel = MethodChannel(engine.dartExecutor.binaryMessenger, "immich/shell")
    channel.setMethodCallHandler { call, result -> result.success(handle(call)) }
    this.channel = channel
  }

  /** The tabs whose root this shell draws itself: the first one, the native grid. */
  val nativeRoots: List<String> get() = listOfNotNull(tabs.firstOrNull()?.id)

  /** Most calls have no reply; `ready` answers with what Dart must not draw. */
  @Suppress("UNCHECKED_CAST")
  private fun handle(call: MethodCall): Any? {
    val args = call.arguments as? Map<String, Any?> ?: emptyMap()
    when (call.method) {
      "ready" -> {
        isReady = true
        tabs = (args["tabs"] as? List<Map<*, *>> ?: emptyList()).mapNotNull(ShellTab::from)
        shellLog("[shell] dart ready, tabs=[${tabs.joinToString(",") { it.id }}] nativeRoots=[${nativeRoots.joinToString(",")}]")
        onTabsDeclared?.invoke(tabs)
        pendingRoute?.let { route ->
          pendingRoute = null
          val settle = pendingSettle ?: {}
          pendingSettle = null
          send(route, settle)
        }
        return mapOf("nativeRoots" to nativeRoots)
      }
      "auth" -> {
        val next = if (args["signedIn"] == true) AuthState.SIGNED_IN else AuthState.SIGNED_OUT
        if (next == authState) return null
        authState = next
        shellLog("[shell] auth signedIn=${next == AuthState.SIGNED_IN}")
      }
      "openViewer" -> onOpenViewer?.invoke(args["session"] as? Int ?: 0, args["index"] as? Int ?: 0)
      "log" -> shellLog("[shell:dart] ${args["text"] ?: "?"}")
      "search" -> {
        (args["placeholder"] as? String)?.let { searchPlaceholder = it }
        (args["text"] as? String)?.let { searchText = it }
      }
      "theme" -> {
        palette = ShellPalette.from(args)
        shellLog("[shell] theme dark=${palette?.dark}")
      }
      "barScroll" -> onBarScroll?.invoke(args["route"] as? String ?: "", (args["progress"] as? Number)?.toFloat() ?: 0f)
      "barCollapsed" -> onBarCollapsed?.invoke(args["route"] as? String ?: "", args["collapsed"] == true)
      "willChange" -> onWillChange?.invoke()
      "sync" -> {
        val frames = (args["routes"] as? List<Map<String, Any?>> ?: emptyList()).map(::frame)
        val tabBars = (args["tabBars"] as? Map<String, Map<String, Any?>> ?: emptyMap()).mapValues { frame(it.value) }
        val tab = args["tab"] as? String ?: ""
        if (args["claimTab"] == true && tab.isNotEmpty() && tab != selectedTab) {
          shellLog("[shell:nav] dart moved to $tab")
          selectedTab = tab
        }
        onSync?.invoke(frames, tabBars, tab)
        ShellEngine.dartIsShowing(args["surface"] as? String ?: "", overlay = args["overlay"] == true)
      }
      else -> shellLog("[shell] unhandled dart call ${call.method}")
    }
    return null
  }

  @Suppress("UNCHECKED_CAST")
  private fun frame(raw: Map<String, Any?>) = MirrorFrame(
    name = raw["name"] as? String ?: "?",
    title = raw["title"] as? String,
    actions = raw["actions"] as? List<Map<String, Any?>> ?: emptyList(),
    hero = raw["hero"] == true,
  )

  // MARK: native → Dart

  fun requestSync() {
    channel?.invokeMethod("resync", null)
  }

  fun requestDartPop(route: String) {
    channel?.invokeMethod("popFromNative", mapOf("name" to route))
  }

  fun barAction(route: String, index: Int, item: Int) {
    channel?.invokeMethod("barAction", mapOf("route" to route, "index" to index, "item" to item))
  }

  fun submitSearch(text: String) {
    shellLog("[shell] search submit ${text.ifEmpty { "(cleared)" }}")
    channel?.invokeMethod("searchSubmitted", mapOf("text" to text))
  }

  fun popToRoot(tab: String) {
    channel?.invokeMethod("popToRoot", mapOf("tab" to tab))
  }

  private var reportedInsets: ShellInsets? = null

  fun report(insets: ShellInsets) {
    if (insets == reportedInsets) return
    reportedInsets = insets
    channel?.invokeMethod(
      "insets",
      mapOf(
        "top" to insets.top.toDouble(),
        "bottom" to insets.bottom.toDouble(),
        "left" to insets.left.toDouble(),
        "right" to insets.right.toDouble(),
      ),
    )
    shellLog("[shell] insets top=${insets.top} bottom=${insets.bottom}")
  }

  fun show(route: String, whenSettled: (String) -> Unit) {
    if (!isReady) {
      pendingRoute = route
      pendingSettle = whenSettled
      return
    }
    send(route, whenSettled)
  }

  private fun send(route: String, whenSettled: (String) -> Unit) {
    shellLog("[shell] show route=$route")
    val channel = channel ?: return whenSettled("")
    channel.invokeMethod("show", mapOf("route" to route), object : MethodChannel.Result {
      override fun success(result: Any?) = whenSettled((result as? Map<*, *>)?.get("surface") as? String ?: "")
      override fun error(code: String, message: String?, details: Any?) = whenSettled("")
      override fun notImplemented() = whenSettled("")
    })
  }
}

/** Flutter's resolved `ColorScheme`, as ARGB ints; missing keys fall back to the Material default. */
class ShellPalette(val dark: Boolean, private val colors: Map<String, Int>) {
  operator fun get(key: String): Int? = colors[key]

  companion object {
    fun from(args: Map<String, Any?>) = ShellPalette(
      dark = args["dark"] == true,
      colors = args.filterKeys { it != "dark" }.mapNotNull { (k, v) -> (v as? Number)?.let { k to it.toInt() } }.toMap(),
    )
  }
}

/** Logical pixels, as Flutter's `MediaQuery` wants them. */
data class ShellInsets(val top: Float, val bottom: Float, val left: Float, val right: Float) {
  val isZero get() = top == 0f && bottom == 0f && left == 0f && right == 0f
}
